kernel/src/hil/screen.rs

```diff
@@ -265,6 +349,6 @@ pub trait ScreenClient {
     /// This is different from `command_complete` as it has to pass back the write buffer
     fn write_complete(&self, buffer: &'static mut [u8], r: Result<(), ErrorCode>);
 
-    /// Some screens need some time to start, this function is called when the screen is ready
-    fn screen_is_ready(&self);
+    /// Some screens need some time to start, this function is called when the screen is ready.
+    fn screen_is_ready(&self, readiness: UpdateType);
```

Comment:
```suggestion
```

I think you are right and we can delete this. Applications should issue a power command that should initialize the screen. Any command sent before that should fail with OFF.

I do not fully understand the `UpdateType`.

Reply:
UpdateType is a consequence of allowing atomic requests. Let's say that the device is able to turn on with a given resolution and color space already configured (for example because the "reset" call has space for them, and setting them after power up takes extra time). The user will want them to be configured as soon as the "ready" signal arrives.

However, we can't unconditionally require this ability from all drivers in the HIL. Another device may initialize in the native resolution no matter what. Then "UpdateType" comes in, allowing the user to know which device was just started.

The first ("atomic") device successfully started and already set the resolution. It returns UptateType::Complete.

The second ("sequential") device started, but ignored the requested resolution. It returns UpdateType::Partial.

I included this atomic mode setting, because it's the same as setting brightness before startup, and presumably costs very little extra complexity.

Reply:
I think I understand somehow your use case. My take in this is that the two things should not be mixed. Powering on the device should just do that and applications should not make any assumption upon resolution and color mode. As soon as the device powers on, applications should request the color mode and resolution, and if required, set them.

I think that adding `UpdateType` adds to much complexity. My suggestion is to keep only one single callback, `command_complete` and state that all commands will fail if the display is power off.

Reply:
This is the same as setting brightness. You do not want to turn on your device and end up blinding the user for a split second because brightness is detached from power. Similarly, you do not want to have flicker because you're changing resolution or color mode.

Sometimes you can't get away with it, but I think it's a reasonable thing to have foreseen in an API.

I think that setting the brightness before power is a very basic thing and I'm going to have to be convinced to remove it, even if it adds complexity, because of how useful it is for user experience.

Reply:
I do not exactly understand how the `UpdateType` can stop the flicker.

I am not convinced that power on and brightness should be mixed. In my experience with working with low end displays, powering them off means they might loose any configuration that was made. From this point of view, an application needs to assume that when powering down a display means it might loose any configurations. If an application wants to turn off the screen without loosing configuration, it should set the brightness to 0.

Moreover, low end displays are not configurable when powered down (they might actually be powered down).

My view on this is the following: 
- applications should consider that displays need to be powered on before they can be configured or used. Any command issues before that may fail and might not be reflected after power up
- drivers should set the brightness of the screens to 0 after powering on (for the simple displays that I have used this means that they do not actually show anything)
- applications should set a brightness above 0 whenever they need to show content to the user (this might mean that applications can draw, but the user does not see anything unless the brightness is set above 0)

Reply:
UpdateType does not stop the flicker. You could drop it and have no flicker on supporting displays.

Powered down displays are not configurable when powered down, and they are not meant to. This is not attempting to change that. Instead, it's attempting to allow changing two properties atomically: brightness and power state.

If my display has a hardware functionality: PowerOnWithBrightness(brightness_level), I don't want to go through PowerOnWithBrightness(1) and then SetBrightness(desired_level), because that causes flicker, and takes extra time. Why not use the interface that's available?

From a hardware designer standpoint, the display may have multiple properties at poweron: resolution, brightness, rotation, pixel format, filter, etc. The hardware may initialize any of them to default, or may allow the user to set any combination inside the PowerOn call. From an API designer standpoint, you don't want to have a separate function for each possible combination. One way to avoid it is to send each property to the driver before requesting power-on, and let the driver choose the right call.

To the user, it looks like "configuring" the display before it's on, but it's not.

> drivers should set the brightness of the screens to 0 after powering on

This breaks if the driver doesn't support 0 while powered on, and also doesn't solve flickering.

> applications should set a brightness above 0 whenever they need to show content to the user 

Which is as soon as the display is on.

As for UpdateType, imagine sending SetPowerWithResolution(resolution). On a display that does not support a given resolution, how do you communicate the resulting state of the display?

Reply:
I understand your point of view. I still believe we need to keep system calls as simple as possible, exactly as we do not know what kind of commands a display can actually take. Functions like PowerOnWithResolution would not work exactly for this reason.

You do have a point about not having to resend all the command again. I suggest deferring these actions to the driver in the following manner.

- functions called between power on and power off are executed immediately
- for functions called after power off, the driver should store the requested command and execute it together with the power on command

In this way we do keep commands simple, while avoiding duplicates.

Reply:
I am currently writing a new improved HIL and TRD (#3055) for the display, so I suggest we add a similar functionality there and exclude the `UpdateType` for this PR.

Reply:
There isn't a separate upcall for partial updates. The upcall carrying UpdateType is the upcall notifying about how power on completed.

Reply:
I agree, I still don't understand the benefit of knowing if the screen has all the properties applied to it as long as it cannot take any commands until it is fully ready.

Reply:
I can drop it for this MR, except for the brightness.

Reply:
I am sorry, but I still don't understand the usefulness of `UpdateType`. Let's assume we have a display that is powered off. The following steps are run:

- the application issues a set_brightness command, the driver memorizes it and will execute it after the power on command
- the application sends a power on command, the driver powers on the display, reports `UpdateType::Incomplete` and sends the set brightness command to the display
- the application receives the upcall with `UpdateType::Incomplete`. This is where I don't get how does this help the application? Any command that it tries to send to the driver will return `BUSY`
- the display finishes setting the brightness, notifies the driver that sends the `UpdateType::Complete` upcall. Now an application can send commands

Reply:
I think we can skip UpdateType and atomic mode changes for this MR, for the sake of not holding up uncontroversial improvements. I'll update it in the coming days.

For an immediate explanation:


-    the application issues a set_brightness command, the driver memorizes it and will execute it after the power on command
-    the application sends a power on command, the driver atomically powers on the display at selected brightness
-    the application receives the upcall with UpdateType::Complete

OR 


-    the application issues a set_brightness command, the driver memorizes it and will execute it after the power on command
-    the application sends a power on command, the driver sequentially powers on the display, and then attempts to set brightness but fails
-    the application receives the upcall with UpdateType::Incomplete

Reply:
This makes things more clear now. The issue that I see is that if there are more parameters than brightness, the application has no idea which one failed and there is not much that it can do (maybe power cycle the display?).

Instead of q quick fix now, I'd rather find a more general approach that allows the application to take an action.

Reply:
I'm sorry, but I think your example needs to be more clear, I personally do not understand this from your code example.

Could you please use code that actually compiles, instead of functions like `ignore_off` which do no actually exist. I think it's important to have clear working examples. Of course, you can skip the registering of the callback function.

Reply:
I've removed this entire train of thought from this proposal. We can consider it when it becomes needed.